README.md
[![](https://flat.badgen.net/npm/v/searchinghost)](https://www.npmjs.com/package/searchinghost)
[![](https://flat.badgen.net/bundlephobia/minzip/searchinghost?label=gzip%20size)](https://bundlephobia.com/result?p=searchinghost)
[![](https://data.jsdelivr.com/v1/package/npm/searchinghost/badge)](https://www.jsdelivr.com/package/npm/searchinghost)


# SeachinGhost

A pure javascript, lightweight & in-browser full-text search plugin for [Ghost](https://ghost.org/) (blog)


## Description

SearchinGhost is a lightweight and extensible search engine dedicated to the Ghost blogging platform. In its core, it uses
the [Ghost Content API](https://ghost.org/docs/api/v3/content/) to load your blog content and the powerful
[FlexSearch](https://github.com/nextapps-de/flexsearch) library to index and execute search queries.

Everything happens in the client browser, it helps us to deliver blazing fast search results and displays them
in real-time to your users (a.k.a "search-as-you-type"). We also take care about minimizing network loads by
relying on the browser `localStorage` only sending requests when necessary.

You blog is in Cyrillic, Chinese, Korean, Greek, Indian or any other non-latin languague? No worries, it is supported, see the [dedicated section](#language-settings).

*__BONUS__: if you like the concept but you would like to install it quickly and easily (in less than 3 minutes, really!), please visit the [SearchinGhostEasy project](https://github.com/gmfmi/searchinghost-easy).*


## Demo

Before diving into installation and configuration, give it a try by yourself with this **[Live Demo](https://gmfmi.github.io/searchinGhost/)**.

On this demo, the searchable content comes from the official Ghost demo API (i.e. https://demo.ghost.io).
Options are set to default so each input word is searched into posts title, tags, excerpt and main content.

For example, search the word "marmalade". It does not exist in any post title, exerpt or tag but it is used one time in the "Down The Rabbit Hole" article, that why you will get it as a result.


## Quick Start

First, update the `default.hbs` file of your theme to include an input field and an output element to display the search results. Then, add a link to SearchinGhost script and initialize it with your own `CONTENT_API_KEY`. To get the content API key, please refer to the official [Ghost documentation](https://ghost.org/docs/api/v3/content/#key).

```html
<input id="search-bar">
<ul id="search-results"></ul>

<script src="https://cdn.jsdelivr.net/npm/searchinghost@1.6.0/dist/searchinghost.min.js"></script>
<script>
    var searchinGhost = new SearchinGhost({
        key: 'CONTENT_API_KEY'
    });
</script>
```

That's it, everything is done! If you need a more fine-grained configuration, please read the next sections.


## Installation

You can install SearchinGhost using various methods, here are the possibilities:

1. **From a Content Delivery Network (CDN)**

This is the easiest and prefered method to install SearchinGhost. Add one of these scripts
into your theme `default.hbs`. We also recommand the use of jsdelivr over unpkg for its
reliability and performance.

```html
<script src="https://cdn.jsdelivr.net/npm/searchinghost@1.6.0/dist/searchinghost.min.js"></script>
<!-- OR -->
<script src="https://unpkg.com/searchinghost@1.6.0/dist/searchinghost.min.js"></script>
```

2. **From source**

If you want to serve SearchinGhost from your own server or include it in your build process,
you can get it from the [release page](https://github.com/gmfmi/searchinGhost/releases) assets or
download the `dist/searchinghost.min.js` file.

3. **From NPM**

Install SearchinGhost as a project dependency.

```shell
$ npm install searchinghost
# OR
$ yarn add searchinghost
```

Then, load it from any Javascript file.

```js
import SearchinGhost from 'searchinghost';
// OR
var SearchinGhost = require('searchinghost');
```


## Configuration

The only mandatory configuration field is `key`. Any other field has a default value and became optional.

SearchinGhost has been designed to work out of the box, this minimal configuration is yet powerful! At
each key stroke, it will search into posts title, tags, excerpt and content. This is the default behaviour
as it seems to be the most common.

```js
// SearchinGhost minimal configuration
var searchinGhost = new SearchinGhost({
    key: '<CONTENT_API_KEY>'
});
```

Nevertheless, a little bit of extra configuration could be worth it to fit your needs.
Let's say you only want search into the `title` and display the `title` and `published_at`
fields for each post found. You could use this configuration:

```js
var searchinGhost = new SearchinGhost({
    key: '<CONTENT_API_KEY>',
    postsFields: ['title', 'url', 'published_at'],
    postsExtraFields: [],
    postsFormats: [],
    indexedFields: ['title'],
    template: function(post) {
        return `<a href="${post.url}">${post.published_at} - ${post.title}</a>`
    }
});
```

SearchinGhost is easily customizable and extensible through its configuration, take your
time to look into each option from the next section.


## Options

### The complete default configuration

```js
{
    key: '',
    url: window.location.origin,
    version: 'v3',
    loadOn: 'focus',
    searchOn: 'keyup',
    limit: 10,
    inputId: ['search-bar'],
    outputId: ['search-results'],
    outputChildsType: 'li',
    postsFields: ['title', 'url', 'excerpt', 'custom_excerpt', 'published_at', 'feature_image'],
    postsExtraFields: ['tags'],
    postsFormats: ['plaintext'],
    indexedFields: ['title', 'string_tags', 'excerpt', 'plaintext'],
    template: function(post) {
        var o = `<a href="${post.url}">`
        if (post.feature_image) o += `<img src="${post.feature_image}">`
        o += '<section>'
        if (post.tags.length > 0) {
            o += `<header>
                    <span class="head-tags">${post.tags[0].name}</span>
                    <span class="head-date">${post.published_at}</span>
                  </header>`
        } else {
            o += `<header>
                    <span class="head-tags">UNKNOWN</span>
                    <span class="head-date">${post.published_at}</span>
                  </header>`
        }
        o += `<h2>${post.title}</h2>`
        o += `</section></a>`
        return o;
    },
    emptyTemplate: function() {},
    customProcessing: function(post) {
        if (post.tags) post.string_tags = post.tags.map(o => o.name).join(' ').toLowerCase();
        return post;
    },
    date: {
        locale: document.documentElement.lang || "en-US",
        options: { year: 'numeric', month: 'short', day: 'numeric' }
    },
    cacheMaxAge: 1800,
    onFetchStart: function() {},
    onFetchEnd: function(posts) {},
    onIndexBuildStart: function() {},
    onIndexBuildEnd: function(index) {},
    onSearchStart: function() {},
    onSearchEnd: function(posts) {},
    indexOptions: {},
    searchOptions: {},
    debug: false
});
```


### All available options

- **key** (string, mandatory)
> The public content API key to get access to blog data.
>
> example: `'22444f78447824223cefc48062'`

- **url** (string)
> The full domaine name of the Ghost API.
>
> example: `'https://demo.ghost.io'`
>
> default: `window.location.origin`

- **version** (string)
> Set the Ghost API version. Work with both `'v2'` and `'v3'`.
>
> default: `'v3'`

- **loadOn** (string)
> Set the library loading strategy. It can be triggered when the HTML page has
> been loaded, on demand when the user click on the search bar or never.
>
> To trigger the search bar initialization by yourself, set this value to `false` (boolean).
> This way, your can call `searchinGhost.loadData()` when the rest of your code is ready.
>
> expected values: `'page'`, `'focus'` or `false`
>
> default: `'focus'`

- **searchOn** (string)
> Choose when the search query should be executed. To search at each
> user key stroke and form submit, use `'keyup'`. To search only when the user submit the
> form via a button or entering the 'enter' key, use `'submit'`. If
> you want to have a complete control of it from your own javascript
> code, use `false` (boolean) and execute the search by yourself using
> `searchinGhost.search("...")`.
>
> expected values: `'keyup'`, `'submit'` or `false`
>
> default: `'keyup'`

- **limit** (number)
> Set the maximum number of posts returned by a search query. Any value between `1`
> and `50` will be lightning-fast and a values lower than `1000` should not
> degrade performance too much. But remember, when the search engine hits this limit
> it stops digging and return the results: the lower, the better.
>
> Even though it is strongly discouraged, set this value to `0` to display all the available results.
>
> default: `10`

- **inputId** (array of string)
> \[Deprecated] Before `v1.6.0`, this field was a `string`, this behaviour has been deprecated.
>
> Your website may have one or several search bars, each of them must have a unique
> HTML `id` attribute. Put each search bar  `id` in this array.
> Do not include '#' in the name.
>
> If you do not need any input field, set the value to `[]` (empty array) and also set `searchOn`
> to `false` (boolean). Then, run a search using `searchinGhost.search("<your query>")`.
>
> default: `['search-bar']`

- **outputId** (array of string)
> \[Deprecated] Before `v1.6.0`, this field was a `string`, this behaviour has been deprecated.
>
> Your website cann use one or more HTML elements to display the search results.
> This array reference all these output element's `id` attribute.
> If any of these elements has already a content, it will be overwitten by
> the search results.
>
> If you are using a JS framework to display the search results, set this value to
> `[]` (empty array). You will get the posts found as the value returned by the
> function `searchinGhost.search("<your query>")`.
>
> default: `['search-results']`

- **outputChildsType** (string)
> \[Deprecated] Before the `v1.6.0`, this fields was a `string`. this has been deprecated.
>
> Each search result is wrapped inside a child element before being added to the
> `outpudId` parent element. The default type is `li` but you can set it to any
> valid HTML element (see [MDN docs](https://developer.mozilla.org/en-US/docs/Web/HTML/Element)).
>
> If you do not want to use a wrapping element to directly append the results of
> `template` and `emptyTemplate` to the output element, set the value to `false` (boolean).
>
> default: `'li'`

- **postsFields** (array of strings)
> An array of all desired posts fields. All these fields will become available
> in the `template` function to display useful posts information.
>
> Refer to the "fields" [official documentation](https://ghost.org/docs/api/v3/content/#fields).
>
> Note: if you use `'custom_excerpt'`, its content will automatically be put in `'excerpt'` to make
> templating easier.
>
> default: `['title', 'url', 'excerpt', 'custom_excerpt', 'published_at', 'feature_image']`

- **postsExtraFields** (array of strings)
> This array allows you to use extra fields like `tags` or `authors`. I personnaly don't know
> why they are not with the other "fields" but the Ghost API is designed this way...
>
> Set its value to `[]` (empty array) to completely disable it.
>
> Refer to the "include" [official documentation](https://ghost.org/docs/api/v3/content/#include).
>
> default: `['tags']`

- **postsFormats** (array of strings)
> This correspond to the "formats" Ghost API which allows use to fetch
> the posts content with HTML or plain text.
>
> Set its value to `[]` (empty array) to completely disable it.
>
> Please refer to the "formats" [official documentation](https://ghost.org/docs/api/v3/content/#formats).
>
> default: `['plaintext']`

- **indexedFields** (array of strings)
> List of indexed fields. The content of all these fields will be searchable.
>
> All value in this list **must** be defined in the posts. Otherwise, the search
> result won't be accurate but the app won't crash! Double check `postsFields`,
> `postsExtraFields` and `postsFormats` values.
>
> *NOTE*: the `'string_tags'` weird field is added in the `customProcessing` option.
> If you want to use tags, this ugly thing is necessary (for now) because FlexSearch
> cannot properly handle arrays. If you do not want/like it, override the `customProcessing`
> to only return `posts` without additionnal modification. If you decide to use tags,
> please also use `'string_tags'` here.
>
> default: `['title', 'string_tags', 'excerpt', 'plaintext']`

- **template** (function)
> Define your own result template. This template will be used for each post found to
> generate the result and appended as child element to the output element. There is
> no templating engine, just a native javascript function using a `post` object as argument.
>
> This template option is lot more powerful than you might expect. We can also think of
> it as a custom processing function called on the search results. For example, if
> you want to do some filtering, return nothing (e.g. `return;`) or an empty string
> (e.g. `return "";`) to discard an item.
>
> Please note the use of **backticks** (e.g. '`') instead of single/double quotes. This
> is required to enable the very useful javascript [variable interpolation](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals).
>
> The available variables are the ones defined in the `postsFields` option.
>
> example:
> ```js
> template: function (post) {
>    return `<a href="${post.url}">#${post.tags} - ${post.published_at} - ${post.title}</a>`
> }
> ```

- **emptyTemplate** (function)
> Define your own result template when there is no result found.
>
> example:
> ```js
> emptyTemplate: function() {
>   return '<p>Sorry, nothing found...</p>'
> }
> ```

- **customProcessing** (function)
> You need to do some extra modification on the posts data fetched from Ghost?
> Use this function to do whatever you need. This function is called on each post,
> executed after the `onFetchEnd()` and before the `onIndexBuildStart()`.
>
> If you want to discard a post, return any JS falsy value (e.g. `null`,
> `undefined`, `false`, `""`, ...).
>
> To easily debug your inputs/outputs, use the `onFetchEnd()` and `onIndexBuildEnd()`
> to display the result with a `console.log()`. If you are a more advanced user, the
> best option is still to use the debugger. Also, do not forget to clean your local
> cache when testing!
>
> *note*: by default, this option is already filled with a helper function to make it
> easier to use the field "tags" in posts. See the `indexedFields` options.
>
> example:
> ```js
> customProcessing: function(post) {
>   post.extra_field = "hello";
>   return post;
> }
> ```

- **date** (object)
> Define the date format fetched from posts.
>
> See the [MDN reference](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/toLocaleDateString#Using_options) to get more information.
>
> example:
> ```js
> date: {
>     locale: "fr-FR",
>     options: { weekday: 'long', year: 'numeric', month: 'long', day: 'numeric' }
> }
> ```

- **cacheMaxAge** (number)
> Set the cache maximum age in seconds. During this amont of time, if an already
> existing index is found in the local storage, it will be loaded without any
> additional HTTP request to confirm its validity. When the cache is purged, the value is reset.
>
> This is espacially useful to save the broadband and network loads of your server. The default
> value is set to half an hour. This value comes from the default user session duration used by
> Google analytics.
>
> default: `1800`

- **onFetchStart** (function)
> Define a callback function before we fetch the data from the Ghost API.
>
> This function takes no argument.
>
> example:
> ```js
> onFetchStart: function() {
>   console.log("before data fetch");
> }
> ```

- **onFetchEnd** (function)
> Define a callback function when the fetch is complete. Even if modifications
> made to `posts` are persisted, we recommand the use of the `customProcessing()`
> function to do so.
>
> The function takes one argument: the array of all posts returned by Ghost itself.
>
> example:
> ```js
> onFetchEnd: function(posts) {
>   console.log("Total posts found on Ghost:", posts.lenght);
> }
> ```

- **onIndexBuildStart** (function)
> Define a callback function before we start building the search index.
>
> The function takes no argument.
>
> example:
> ```js
> onIndexBuildStart: function() {
>   console.log("before building the index");
> }
> ```

- **onIndexBuildEnd** (function)
> Define a callback function when the search index build is complete.
>
> The function takes one argument: the build FlexSearch index object.
>
> example:
> ```js
> onIndexBuildEnd: function(index) {
>   console.log("index built:", index);
> }
> ```

- **onSearchStart** (function)
> Define a callback function before starting to execute the search query. For
> instance, it could be used to hide the results HTML element while waiting for the
> `onSearchEnd` completion or add any fancy transition effects. But in most cases,

CHANGELOG.md
# CHANGELOG

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).


## [Unreleased]
(empty)


## [1.6.2] - 2020-11-08
### Fixed
- Fix the backward compability issue of `inputId` and `outputId` of release `v1.6.0` and `v1.6.1`.


## [1.6.1] - 2020-11-08
### Fixed
- Detect "deleted" and "unpublished" posts to trigger a cache invalidation


## [1.6.0] - 2020-08-22
### Changed
- The fields `inputId` and `outputId` now support array values. Very useful for multi search bars and/or
  multi result elements setup. Configuration backward compability is preserved.


## [1.5.0] - 2020-08-05
### Added
- A new `searchOptions` field to fine tune Flexsearch searches.

### Changed
- Better info and error console logs, the error object is also printed out.


## [1.4.0] - 2020-07-05
### Changed
- The `outputId` becomes optional, set it value to `false` to disable it.
- Replace all `throw` statement by warn logs.
- The `search()` method now returns the posts found.


## [1.3.3] - 2020-05-28
### Changed
- The `outputChildsType` wrapping element becomes optional, set it value to `false` to disable it.
- Use `false`(boolean) value instead of `'none'` for the fields "loadOn" and "searchOn".
  The `'none'` value continues to work but has been deprecated.


## [1.3.2] - 2020-05-26
### Fixed
- Enable library import using yarn/NPM.


## [1.3.1] - 2020-05-23
### Added
- Ability to set `inputId` to `false` when no input field is needed.


## [1.3.0] - 2020-05-23
### Added
- Search engine configuration exposed to user.
- Non-latin languages support.

### Changed
- Remove the Ghost API library dependency, use the native `fetch()` instead. The bundle size is reduced by 50%.


## [1.2.0] - 2020-05-14
### Added
- Self documented methods within the source code (JSdoc).

### Changed
- The inner `loadResources()` method has been renamed to `loadData()`.
- Refactor some method and variable names to give them more sens.

### Fixed
- When `searchOn` set to `none`, properly disable form submission.


## [1.1.0] - 2020-05-10
### Added
- Discard post item when returning falsy value from the `customProcessing()` callback function.

### Changed
- Default format date based on the HTML page `lang` attribute.


## [1.0.0] - 2020-05-08

The first official and production-ready release of SearchinGhost!


src/searchinghost.js
import FlexSearch from 'flexsearch';

export default class SearchinGhost {

    /**
     * Constructor and entry point of the library
     * @param {Document} args 
     */
    constructor(args) {
        this.config = {
            url: window.location.origin,
            key: '',
            version: 'v3',
            loadOn: 'focus',
            searchOn: 'keyup',
            limit: 10,
            inputId: ['search-bar'],
            outputId: ['search-results'],
            outputChildsType: 'li',
            postsFields: ['title', 'url', 'excerpt', 'custom_excerpt', 'published_at', 'feature_image'],
            postsExtraFields: ['tags'],
            postsFormats: ['plaintext'],
            indexedFields: ['title', 'string_tags', 'excerpt', 'plaintext'],
            template: function(post) {
                var o = `<a href="${post.url}">`
                if (post.feature_image) o += `<img src="${post.feature_image}">`
                o += '<section>'
                if (post.tags.length > 0) {
                    o += `<header>
                            <span class="head-tags">${post.tags[0].name}</span>
                            <span class="head-date">${post.published_at}</span>
                          </header>`
                } else {
                    o += `<header>
                            <span class="head-tags">UNKNOWN</span>
                            <span class="head-date">${post.published_at}</span>
                          </header>`
                }
                o += `<h2>${post.title}</h2>`
                o += `</section></a>`
                return o;
            },
            emptyTemplate: function() {},
            customProcessing: function(post) {
                if (post.tags) post.string_tags = post.tags.map(o => o.name).join(' ').toLowerCase();
                return post;
            },
            date: {
                locale: document.documentElement.lang || "en-US",
                options: { year: 'numeric', month: 'short', day: 'numeric' }
            },
            cacheMaxAge: 1800,
            onFetchStart: function() {},
            onFetchEnd: function(posts) {},
            onIndexBuildStart: function() {},
            onIndexBuildEnd: function(index) {},
            onSearchStart: function() {},
            onSearchEnd: function(posts) {},
            indexOptions: {},
            searchOptions: {},
            debug: false
        }
        
        this.dataLoaded = false;  // flag to ensure data are properly loaded
        this.postsCount = 0;      // keep track of posts ID, must be numeric
        this.storage = this.getLocalStorageOption();

        this.initConfig(args);
        this.triggerDataLoad();
    }

    /**
     * Apply the user configuration and initialize important variables
     * @param {Document} args 
     */
    initConfig(args) {
        for (let [key, value] of Object.entries(args)) {
            this.config[key] = value;
        }

        // ensure config backward compatilibity of <1.5.0
        if (!Array.isArray(this.config.inputId)) this.config.inputId = [this.config.inputId];
        if (!Array.isArray(this.config.outputId)) this.config.outputId = [this.config.outputId];

        // Inject the "limit" arg within the final searchOptions
        this.config.searchOptions.limit = this.config.limit;

        // Ensure 'updated_at' will be fetched, needed for the local storage logic
        this.originalPostsFields = this.config.postsFields;
        if (!this.config.postsFields.includes('updated_at')) {
            this.config.postsFields.push('updated_at');
        }

        if (this.config.inputId && this.config.inputId.length > 0) {
            this.searchBarEls = [];
            this.config.inputId.forEach(id => {
                let searchBar = document.getElementById(id);
                if (searchBar) {
                    this.searchBarEls.push(searchBar);
                    this.addSearchListeners(searchBar);
                } else {
                    this.error(`Enable to find the input element #${id}, please check your configuration`);
                }
            });
        }

        if (this.config.outputId && this.config.outputId.length > 0) {
            this.searchResultEls = [];
            this.config.outputId.forEach(id => {
                let searchResult = document.getElementById(id);
                if (searchResult) {
                    this.searchResultEls.push(searchResult)
                } else {
                    this.error(`Enable to find the output element #${id}, please check your configuration`);
                }
            });
        }

        this.index = this.getNewSearchIndex();
    }

    /**
     * Set the search input bar and form event listeners to trigger
     * further searches
     */
    addSearchListeners(searchBarEl) {
        // In any case, prevent the input form from being submitted
        let searchForm = searchBarEl.closest('form');
        if (searchForm) searchForm.addEventListener("submit", (ev) => { ev.preventDefault(); });

        switch(this.config.searchOn) {
        case 'keyup':
            searchBarEl.addEventListener("keyup", () => {
                let inputQuery = searchBarEl.value.toLowerCase();
                this.search(inputQuery);
            });
            break;
        case 'submit':
            searchForm.addEventListener("submit", () => {
                let inputQuery = searchBarEl.value.toLowerCase();
                this.search(inputQuery);
            });
            break;
        case false:
        case 'none':
            // do nothing
            break;
        default:
            this.error(`Unknown "searchOn" option: '${this.config.searchOn}'`);
        }
    }

    /**
     * Set triggers to load the posts data when ready
     */
    triggerDataLoad() {
        switch(this.config.loadOn) {
        case 'focus':
            this.searchBarEls.forEach(searchBarEl => {
                searchBarEl.addEventListener('focus', () => {
                    this.loadData();
                });
            })
            break;
        case 'page':
            window.addEventListener('load', () => {
                this.loadData();
            });
            break;
        case false:
        case 'none':
            // do nothing
            break;
        default:
            this.error(`Unknown "loadOn" option: '${this.config.loadOn}'`);
        }
    }

    /**
     * Actually load the data into a searchable index.
     * When this method is completed, we are ready to launch search queries.
     */
    loadData() {
        if (this.dataLoaded) return;

        if (!this.storage) {
            this.log("No local storage available, switch to degraded mode");
            this.fetch();
            return;
        }

        let storedIndex = this.storage.getItem("SearchinGhost_index");
        if (storedIndex) {
            this.log("Found an index stored locally, loads it");
            this.config.onIndexBuildStart();
            this.index.import(storedIndex);
            this.dataLoaded = true;
            this.config.onIndexBuildEnd(this.index);
            this.validateCache();
        } else {
            this.log("No already stored index found");
            this.fetch();
        }
    }

    /**
     * Ensure stored data are up to date.
     */
    validateCache() {
        let cacheInfoString = this.storage.getItem("SearchinGhost_cache_info");
        if (!cacheInfoString) {
            this.log("No cache info local object found");
            this.fetch();
            return;
        }

        let cacheInfo = JSON.parse(cacheInfoString);

        const lastUpdate = new Date(cacheInfo.lastCacheCheck);
        const elapsedTime = Math.round((new Date() - lastUpdate) / 1000);
        if (elapsedTime < this.config.cacheMaxAge) {
            this.log(`Skip cache refreshing, updated less than ${this.config.cacheMaxAge}s ago (${elapsedTime}s)`);
            return;
        }

        const browseOptions = {
            limit: 1,
            fields: ['updated_at'],
            order: 'updated_at DESC'
        };
        const lastUpdatedPostUrl = this.buildUrl(browseOptions);
        fetch(lastUpdatedPostUrl)
            .then(function(response) {
                return response.json();
            })
            .then((jsonResponse) => {
                const lastestPostUpdatedAt = jsonResponse.posts[0].updated_at;
                const totalPosts = jsonResponse.meta.pagination.total;

                if (lastestPostUpdatedAt !== cacheInfo.lastestPostUpdatedAt) {
                    this.log("Posts update found, purge outdated local cache");
                    this.fetch();
                } else if (totalPosts < cacheInfo.totalPosts) {
                    this.log("Deleted or unpublished posts found, purge outdated local cache")
                    this.fetch();
                } else {
                    this.log("Local cached data up to date");
                    cacheInfo.lastCacheCheck = new Date().toISOString();
                    this.storage.setItem("SearchinGhost_cache_info", JSON.stringify(cacheInfo));
                }
            }).catch((error) => {
                console.error("Unable to fetch the latest post information to check cache state", error);
            });
    }

    /**
     * Fetch, format and store posts data from Ghost.
     */
    fetch() {
        this.log("Fetching data from Ghost API");
        this.config.onFetchStart();

        const browseOptions = {
            limit: 'all',
            fields: this.config.postsFields,
            order: 'updated_at DESC'
        }
        if (this.config.postsExtraFields.length > 0) browseOptions.include = this.config.postsExtraFields;
        if (this.config.postsFormats.length > 0) browseOptions.formats = this.config.postsFormats;

        const allPostsUrl = this.buildUrl(browseOptions);
        fetch(allPostsUrl)
            .then(function(response) {
                return response.json();
            })
            .then((jsonResponse) => {
                const posts = jsonResponse.posts;
                this.config.onFetchEnd(posts);
                this.config.onIndexBuildStart();
                
                this.index = this.getNewSearchIndex();
                posts.forEach((post) => {
                    let formattedPost = this.format(post);
                    if (formattedPost) this.index.add(formattedPost);
                });

                this.dataLoaded = true;
                this.config.onIndexBuildEnd(this.index);

                if (this.storage) {
                    const cacheInfo = {
                        lastCacheCheck: new Date().toISOString(),
                        lastestPostUpdatedAt: posts[0].updated_at,
                        totalPosts: jsonResponse.meta.pagination.total
                    }
                    this.storage.setItem("SearchinGhost_index", this.index.export());
                    this.storage.setItem("SearchinGhost_cache_info", JSON.stringify(cacheInfo));
                }

                this.log("Search index build complete");
            })
            .catch((error) => {
                this.error("Unable to fetch Ghost data.\n", error);
            });
    }

    /**
     * Format a post document before being indexed.
     * @param {Document} post 
     * @return {Document} The formatted post
     */
    format(post) {
        // Need to use a numeric ID to improve performance & disk space
        post.id = this.postsCount++;

        // display date using 'locale' format
        post.published_at = this.prettyDate(post.published_at);

        // only used to keep track of the last fetch time,
        // remove it before indexing BUT only if not wanted by the user
        if (!this.originalPostsFields.includes('updated_at')) {
            delete post.updated_at;
        }

        if (post.custom_excerpt) {
            post.excerpt = post.custom_excerpt;
            delete post.custom_excerpt;
        }

        post = this.config.customProcessing(post);

        return post;
    }

    /**
     * Execute a search query.
     * @param {string} inputQuery 
     */
    search(inputQuery) {
        this.loadData();

        this.config.onSearchStart();

        let postsFound = this.index.search(inputQuery, this.config.searchOptions);

        if (this.searchResultEls && this.searchResultEls.length > 0) this.display(postsFound);

        this.config.onSearchEnd(postsFound);
        return postsFound;
    }

    /**
     * Display the results as HTML into the configured DOM output element.
     * @param {Document[]} posts 
     */
    display(posts) {
        this.searchResultEls.forEach(resultEl => {
            resultEl.innerHTML = '';
        });
        
        if (posts.length < 1) {
            this.insertTemplate(this.config.emptyTemplate());
        } else {
            posts.forEach(post => {
                this.insertTemplate(this.config.template(post));
            });
        }
    }

    /**
     * Insert the HTML generated by the template into the DOM results output element.
     * If a falsy value is returned by the template, do not apply any update.
     * @param {*} generatedHtml HTML node element or HTML string
     */
    insertTemplate(generatedHtml) {
        if (generatedHtml) {
            this.searchResultEls.forEach(resultEl => {
                if (this.config.outputChildsType) {
                    let child = document.createElement(this.config.outputChildsType);
                    child.classList.add(`${resultEl.id}-item`);
                    child.innerHTML = generatedHtml;
                    resultEl.appendChild(child);
                } else {
                    resultEl.insertAdjacentHTML('beforeend', generatedHtml);
                }
            });
        }
    }

    /**
     * Get a new instance of FlexSearch.
     * @return {FlexSearch} The instance of FlexSearch.
     */
    getNewSearchIndex() {
        const indexConfig = {
            doc: {
                id: "id",
                field: this.config.indexedFields
            },
            encode: "simple",
            tokenize: "forward",
            threshold: 0,
            resolution: 4,
            depth: 0
        }

        for (let [key, value] of Object.entries(this.config.indexOptions)) {
            indexConfig[key] = value;
        }
        
        return new FlexSearch(indexConfig);
    }

    /**
     * Build the final Ghost API URL resources based on options.
     * @param {Document} options the Ghost API browse options
     * @return {string} the url
     */
    buildUrl(options) {
        let url = `${this.config.url}/ghost/api/${this.config.version}/content/posts/?key=${this.config.key}`;
        for (const [key, value] of Object.entries(options)) {
            url += `&${key}=${value}`;
        }
        return encodeURI(url);
    }

    /**
     * Get the date in the locale expected format.
     * @param {string} date 
     * @return {string} The formatted date
     */
    prettyDate(date) {
        let d = new Date(date);
        return d.toLocaleDateString(this.config.date.locale, this.config.date.options);
    }

    /**
     * Safely get the local storage object if available.
     * If the user browser disabled it, get `undefined` instead.
     * @return {Storage} The storage object or `undefined`
     */
    getLocalStorageOption() {
        try {
            window.localStorage.setItem('storage-availability-test', '');
            window.localStorage.removeItem('storage-availability-test');
            return window.localStorage;
        } catch(err) {
            return undefined;
        }
    }

    /**
     * Simple logging function.
     * Output logs only if `debug` is set to `true`.
     * @param {string} str the text to output
     * @param {*} obj optional object to output
     */
    log(str, obj) {
        if (this.config.debug) obj ? console.log(str, obj) : console.log(str);
    }

    /**
     * Simple "error" level logging function.
     * @param {string} str the text to output
     * @param {*} obj optional object to output
     */
    error(str, obj) {
        obj ? console.error(str, obj) : console.error(str);
    }
}
